contracts/pcv/utils/ERC20Allocator.sol

```diff
@@ -0,0 +1,454 @@
+    }
+
+    /// ----------- PURE & VIEW Only APIs -----------
+
+    /// @notice returns the target balance for a given PSM
+    function targetBalance(address psm) public view returns (uint256) {
+        return allPSMs[psm].targetBalance;
+    }
+
+    /// @notice function to get the adjusted amount out
+    /// @param amountToDrip the amount to adjust
+    /// @param decimalsNormalizer the amount of decimals to adjust amount by
+    function getAdjustedAmount(uint256 amountToDrip, int8 decimalsNormalizer)
+        public
+        pure
+        returns (uint256 adjustedAmountToDrip)
+    {
+        if (decimalsNormalizer == 0) {
+            adjustedAmountToDrip = amountToDrip;
+        } else if (decimalsNormalizer < 0) {
+            uint256 scalingFactor = 10**(-1 * decimalsNormalizer).toUint256();
+            adjustedAmountToDrip = amountToDrip / scalingFactor;
+        } else {
+            uint256 scalingFactor = 10**decimalsNormalizer.toUint256();
+            adjustedAmountToDrip = amountToDrip * scalingFactor;
+        }
+    }
+
+    /// @notice return the amount that can be skimmed off a given PSM
+    /// @param pcvDeposit pcv deposit whose corresponding psm will have skim amount checked
+    /// returns amount that can be skimmed, adjusted amount to skim and target to send proceeds
+    function getSkimDetails(address pcvDeposit)
+        public
+        view
+        returns (uint256 amountToSkim, uint256 adjustedAmountToSkim)
+    {
+        address psm = pcvDepositToPSM[pcvDeposit];
+        PSMInfo memory toSkim = allPSMs[psm];
+
+        address token = toSkim.token;
+        amountToSkim = IERC20(token).balanceOf(psm) - toSkim.targetBalance;
+
+        /// adjust amount to skim based on the decimals normalizer to replenish buffer
+        adjustedAmountToSkim = getAdjustedAmount(
+            amountToSkim,
+            toSkim.decimalsNormalizer
+        );
+    }
+
+    /// @notice return the amount that can be dripped to a given PSM
+    /// @param psm peg stability module to check drip amount on
+    /// @param pcvDeposit pcv deposit to drip from
+    /// returns amount that can be dripped, adjusted amount to drip and target
+    function getDripDetails(address psm, PCVDeposit pcvDeposit)
+        public
+        view
+        returns (uint256 amountToDrip, uint256 adjustedAmountToDrip)
+    {
+        PSMInfo memory toDrip = allPSMs[psm];
+
+        /// direct balanceOf call is cheaper than calling balance on psm
+        uint256 targetBalanceDelta = toDrip.targetBalance -
+            IERC20(toDrip.token).balanceOf(psm);
+
+        /// drip min between target drip amount and pcv deposit being pulled from
+        /// to prevent edge cases when a venue runs out of liquidity
+        /// only drip the lowest between amount and the buffer,
+        /// as dripping more than the buffer will result in
+        amountToDrip = Math.min(
+            Math.min(targetBalanceDelta, pcvDeposit.balance()),
+            buffer()
```

Comment:
this needs to be decimal normalized in the opposite direction of usual normalization